### PR TITLE
wait until handshaking connections have terminated when closing server

### PR DESCRIPTION
**Pull request by marten-seemann**
This should resolve the painful race condition we're seeing in the integration tests

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -713,8 +716,9 @@ func (s *baseServer) handleInitialImpl(p receivedPacket, hdr *wire.Header) error
 		delete(s.zeroRTTQueues, hdr.DestConnectionID)
 	}
 
-	go conn.run()
+	s.handshakingCount.Add(1)
 	go func() {
```
Comment (sukunrt) [inline server.go:720]: This goroutine can push a connection on the conn queue right when the server is closing. Is that a problem?

**Review by sukunrt — APPROVED** [review]

File: server.go (hunk 1 above)
Comment (sukunrt) [reply server.go:720]: Isn't there a race here? 
```Accept ...
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case conn := <-s.connQueue:
		return conn, nil
	case <-s.errorChan:
		return nil, s.closeErr
	}
```

Comment (sukunrt) [reply server.go:720]: connQueue has a connection and errorChan is closed.

Comment (marten-seemann) [reply server.go:720]: That is true. It's independent from this PR though. I'll open an issue.

Comment (marten-seemann) [reply server.go:720]: #4744

**Pull request merged**